pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -259,6 +261,20 @@ func (configgen *ConfigGeneratorImpl) buildGatewayHTTPRouteConfig(node *model.Pr
 				continue
 			}
 
+			// Make sure we can obtain services which are visible to this virtualService.
+			nameToServiceMap, exist := namespaceAndHostname[virtualService.Namespace]
+			if !exist {
+				services := push.Services(&model.Proxy{
+					ConfigNamespace: virtualService.Namespace,
+				})
+
+				nameToServiceMap = map[host.Name]*model.Service{}
+				for _, service := range services {
```

Comment:
concerned about the efficiency here

Reply:
+1. I have the `namespaceAndHostname` map in which services are indexed by namespace then hostname. These code only call once in every namespace of vs. Any better idea？

Reply:
This is O(services), iterating over all services. The map doesn't really change that much.

We already have services indexed by namespace, see HostnameAndNamespace. I think you can just do something like:

```go
svc, f := HostnameAndNamespace[hostname][namespace]
if !f {
  // Didn't find match for our namespace, pick an arbitrary one
  svc := pickFrom(HostnameAndNamespace[hostname]) # pick need to sort it or something to get consistent ordering?
}
nameToServiceMap[hostname] = svc
```

Reply:
My first commit is like this idea. Code is revert to same with before.